iree/hal/cuda/cuda_device.c

```diff
@@ -229,19 +262,20 @@ static iree_status_t iree_hal_cuda_device_queue_submit(
     iree_hal_queue_affinity_t queue_affinity, iree_host_size_t batch_count,
     const iree_hal_submission_batch_t* batches) {
   iree_hal_cuda_device_t* device = iree_hal_cuda_device_cast(base_device);
+  iree_hal_command_buffer_t* stream_command_buffer;
+  iree_hal_cuda_stream_command_buffer_allocate(&device->context_wrapper,
```

Comment:
This leaks memory and doesn't check the status, but more importantly shouldn't be allocating memory on each submit. Since this is a stateless transient instance you can either stack allocate it (by moving the struct from the .c to the .h and using an _initialize method instead) or create this once on device creation and cache it on the iree_hal_cuda_device_t.

If you do decide to keep it allocating (which I don't think it needs to be) then you should rename it to _create - the convention used is that _allocate requires a _free and has no reference counting, and _create requires a _destroy and does have reference counting. Since the command buffer is a reference counted type it needs _create and _destroy (unless as above you make it stack allocated, in which case it just needs _initialize/_deinitialize).

Reply:
I'm not sure how it should be initialized, is there any reference? I tried to create it through device creation and cache it on iree_hal_cuda_device_t, but it misses command_categories. Now I destroyed the command buffer after the loop, is that okay?